onnx/checker.cc

```diff
@@ -0,0 +1,247 @@
+#include "onnx/checker.h"
+
+#include "onnx/defs/schema.h"
+#include "onnx/utils.h"
+
+#include <unordered_set>
+
+namespace onnx {
+namespace checker {
+
+#define fail(...) throw ValidationError(onnx::MakeString(__VA_ARGS__));
+
+#define enforce_has_field(proto, field)                                    \
+  if (!proto.has_##field()) {                                              \
+    fail("Field '", #field, "' of ", #proto, " is required but missing."); \
+  }
+
+#define enforce_has_repeated_field(proto, field)                    \
+  if (!proto.field##_size()) {                                      \
+    fail("Repeated Field '", #field, "' is required but missing."); \
+  }
+
+#define enforce_non_empty_field(proto, field)                                  \
+  if (proto.field().empty()) {                                                 \
+    fail("Field '", #field, "' of ", #proto, " is required to be non-empty."); \
+  }
+
+void check_value_info(const ValueInfoProto& value_info, int ir_version) {
+  enforce_non_empty_field(value_info, name);
+  enforce_has_field(value_info, type);
+  const auto value_case = value_info.type().value_case();
+  switch (value_case) {
+    case TypeProto::kTensorType: {
+      const auto& type = value_info.type().tensor_type();
+      enforce_has_field(type, elem_type);
+      enforce_has_field(type, shape);
+    } break;
+    case TypeProto::kSparseTensorType: {
+      const auto& type = value_info.type().sparse_tensor_type();
+      enforce_has_field(type, elem_type);
+      enforce_has_field(type, shape);
+    } break;
+    default:
+      fail("Unrecognized type value case: ", value_case);
+  }
+}
+
+void check_tensor(const TensorProto& tensor, int ir_version) {
+  enforce_has_field(tensor, data_type);
+  if (tensor.data_type() == TensorProto::UNDEFINED) {
+    fail("setting data_type field to UNDEFINED is not allowed");
+  }
+
+  enforce_has_repeated_field(tensor, dims);
+
+  int num_value_fields = 0;
+
+  const char* value_field = nullptr;
+
+#define check_data_field(field)             \
+  bool has_##field = tensor.field().size(); \
+  if (has_##field) {                        \
+    ++num_value_fields;                     \
+    value_field = #field;                   \
+  }
+
+  check_data_field(float_data);
+  check_data_field(int32_data);
+  check_data_field(string_data);
+  check_data_field(int64_data);
+  check_data_field(raw_data);
+  check_data_field(double_data);
+  check_data_field(uint64_data);
+
+#undef check_data_field
+
+  if (num_value_fields != 1) {
+    fail("TensorProto should contain one and only one value field.");
+  }
+  if (has_raw_data) {
+    if (tensor.data_type() == TensorProto::STRING) {
+      fail("STRING data should not be stored in raw_data field");
+    }
+  } else {
```

Comment:
nit: put `return` here - it'd be more readable

Reply:
will fix it